src/mod/endpoints/mod_gsmopen/gsmlib/gsmlib-1.10-patched-13ubuntu/debian/gsmsiexfer.1
.\" -*- eval: (nroff-mode) -*-
.de TQ
.br
.ns
.TP \\$1
..
.\" Like TP, but if specified indent is more than half
.\" the current line-length - indent, use the default indent.
.de Tp
.ie \\n(.$=0:((0\\$1)*2u>(\\n(.lu-\\n(.iu)) .TP
.el .TP "\\$1"
..
.TH GSMSIEXFER 1 "" "gsmsiexfer"
.SH NAME
gsmsiexfer \-  Siemens ME file transfer program for Siemens phones S25, S35, S45, ME45, SL45
.SH SYNOPSIS
.B gsmsiexfer \-\-help
.PP
.SH DESCRIPTION
\fIgsmsiexer\fP comes with no man page.  Try gsmsiexfer \-\-help, or
read the source.
.PP
.SH "SEE ALSO"
.BR gsminfo(7),
.BR gsmctl(1),
.BR gsmsendsms(1),
.BR gsmsmsd(8),
.BR gsmsmsstore(1).



ChangeLog
 * libZRTP SDK library, implements the ZRTP secure VoIP protocol.
 * Copyright (c) 2006-2009 Philip R. Zimmermann.  All rights reserved.
 * Contact: http://philzimmermann.com
 * For licensing and other legal details, see the file zrtp_legal.c.

!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
SINCE LIBZRTP v0.80 CHANGELOG IS A PART OF HTML DOCUMENTATION. 
https://developers.zfoneproject.com/libzrtp/wiki/LibzrtpChangeLog
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!


libzrtp 0.7.1											18.11.2008
--------------------------------------------------------------------------------
1.	Fixed bug with hardcoded AES128 cipher for generating SRTP keys. In this 
	version is selected according to ZRTP discovery.
	
2.	Added initialization/deinitalization functions to zrtp helper functions.

libzrtp 0.7.0											04.11.2008
--------------------------------------------------------------------------------
1.	Changes in libzrtp sources tree.

2.	Improvements in libzrtp initialization routine:
	- all global zrtp options were combined in zrtp_config_t structure;
	- zrtp_init() allocates memory for zrtp global context;
	- zrtp_config_defaults() 

3.	Improvements for Passive/Active mode support.
	a) A passive endpoint never sends a commit message, period. Also, it 
	declares itself as a passive endpoint by setting the P flag it its own
	Hello message;
	b) A active endpoint does not send a commit to a passive endpoint, which it 
	recognizes by detecting the P flag;
	c) A passive phone, if acting as a SIP initiator (meaning it initiated the 
	call), rejects all commit packets from everyone;
	d) A passive phone rejects all commit messages from a PBX, which is easily 
	recognized by the M flag.
	Passive mode support is built into the library logic and will be used
	automatically if the developer specifies signaling role by setting 
	is_initiator flag in zrtp_init_session_ctx().

4.	Improvements in ZRTP feedback interface and system-dependent functions.
    There are two types of interface functions in libzrtp: system dependent API
	and realization of helper functions and events. System dependent API in
	defined in zrtp_iface_system.h as set of extern functions. System
	functions are already implemented for several basic platforms in
	zrtp_iface.sys.c libzrtp feedback and helper functions were re-factored
	and implemented as set of callbacks. If the developer doesn't want to
	handle one or another event it may just leave necessary pointer empty.
	See zrtp_init() and zrtp_callback_t, zrtp_iface.h for more details.

5. 	ZRTP configuration approach was improved: zrtp_config_xxx.h contains
	adjustments for necessary target platform. libzrtp contains default
	configs for Linux, OS X, Windows, Window CE and Symbian platforms.
	All ZRTP protocol and behavior related adjustments are collected in
	zrtp_config_user.h. Edit this file to configure libzrtp for your
	needs.
	
6.	Implemented new functions in protocol according to the Internet Draft v 10.

7.	Improved realization of built-in libzrtp scheduler. Fixed bug with crashing
	on performing delay call when zrtp session have been already deleted.
	
8. 	Logging function was improved. Use ZRTP_LOG macro to print log messages. See
	zrtp_log.h for more information.


libzrtp 0.6.8											03.09.2008
--------------------------------------------------------------------------------
ZRTP
1.	Fixed bug with incorrect maximum value for T1 retry interval timer which 
    increased delay between LOOKING_FOR_ZRTP and NO_ZRTP_SUPPORT states. 
    Internal fix, no API changes required;
2.	Implemented version negotiation according to the latest specification.
	libzrtp v 0.6.8 supports ZRTP v0.90 only. No changes required in 
	applications that use the SDK.
3. 	Compilation flag WITH_ZFONE was removed. The developer, who wants to use
	built-in ZRTP cache, has to set name of the ZRTP cache explicitly,
	implementing  zrtp_get_cache_path() function.
4. 	New libzrtp licensing scheme was implemented. It allows the licensing policy 
    to be changed at run time.  See zrtp_license_mode_t doc. for more details. 
    Affected API - zrtp_init().
5. 	ZRTP Protocol version was changed to 0.90 according to ZRTP Internet Draft.
6. 	Some changes in Linux config files: surplus configuration flags were removed
	from ./cfg.XXX templates. 
7.	Added experimental ZRTP messages retries scheduler for slow channels. As 
	ex ample for GSM CSD channel with average bandwidth 6Kb/s. To use this
	option build library with BUILD_FOR_CSD flag.


libzrtp 0.6.6											27.06.2008
--------------------------------------------------------------------------------
ZRTP
1. Small bug was fixed in S0 calculation: when RS1 is corrupted the library uses
   RS2 instead;
2. Some changes in Makefile and building process: unused header were eliminated 	
   from the installation process.
3. -DBUILD_WITH_ZRTP_MUTEXES was replaced with --enable_mutexes option passed to 
   ./configure script. This change allows not to specify any libzrtp compilation 
   flags during user application compilation. --enable-mutexes adds    
   BUILD_ZRTP_MUTEXES definition to the ./config/zrtp_unix_config.h so if you 
   build libzrtp on other platforms - define this flag manually (windows 
   configuration file already includes this option).
4. Clean-up in .h and .c comments was made.


libzrtp 0.6.5											04.06.2008
--------------------------------------------------------------------------------
ZRTP
1. New names for: other_secret - pbxs; srtps - auxs. In bits and secrets storages;
2. RS2 secret was eliminated form DH s0 calculation;
3. Protocol version number was increased to 0.85


libzrtp 0.6.4											19.05.2008
--------------------------------------------------------------------------------
ZRTP
1. According to the new version of the Internet Draft Signaling shared secret was
   removed from the protocol and from the sources. It was not used by interface

src/mod/endpoints/mod_gsmopen/gsmlib/gsmlib-1.10-patched-12ubuntu1/ext/gsmsiexfer.cc
// *************************************************************************
// * GSM TA/ME library
// *
// * File:    gsmsiexfer.cc
// *
// * Purpose: Siemens ME file transfer program
// *
// * Author:  Christian W. Zuckschwerdt  <zany@triq.net>
// *
// * Created: 2001-12-16
// *************************************************************************

#ifdef HAVE_CONFIG_H
#include <gsm_config.h>
#endif
#include <gsmlib/gsm_nls.h>
#include <string>
#ifdef WIN32
#include <gsmlib/gsm_win32_serial.h>
#else
#include <gsmlib/gsm_unix_serial.h>
#include <unistd.h>
#endif
#if defined(HAVE_GETOPT_LONG) || defined(WIN32)
#include <getopt.h>
#endif
#include <stdio.h>
#include <stdlib.h>
#include <errno.h>
#include <gsmlib/gsm_me_ta.h>
#include <gsm_sie_me.h>
#include <gsmlib/gsm_util.h>
#include <gsmlib/gsm_sorted_phonebook.h>
#include <iostream>

using namespace std;
using namespace gsmlib;

#ifdef HAVE_GETOPT_LONG
static struct option longOpts[] =
{
  {"xonxoff", no_argument, (int*)NULL, 'X'},
  {"init", required_argument, (int*)NULL, 'I'},
  {"destination", required_argument, (int*)NULL, 'd'},
  {"source", required_argument, (int*)NULL, 's'},
  {"baudrate", required_argument, (int*)NULL, 'b'},
  {"type", required_argument, (int*)NULL, 't'},
  {"subtype", required_argument, (int*)NULL, 'i'},
  {"help", no_argument, (int*)NULL, 'h'},
  {"version", no_argument, (int*)NULL, 'v'},
  {"verbose", no_argument, (int*)NULL, 'V'},
  {(char*)NULL, 0, (int*)NULL, 0}
};
#else
#define getopt_long(argc, argv, options, longopts, indexptr) \
  getopt(argc, argv, options)
#endif

// I f*ck up this file IO thing.

// read binary object from stdin
BinaryObject readBinaryFile(istream &ifs, string filename)
{
  size_t size = 10000; // Bad coder, no biscuits!
  BinaryObject bnr;
  bnr._data = new unsigned char[size];
  ifs.read((char*)bnr._data, size);
  bnr._size = ifs.gcount();
  return bnr;
}

// read binary object from file
BinaryObject readFile(string filename)
{
  // open the file
  ifstream ifs(filename.c_str());
  if (ifs.bad())
    throw GsmException(stringPrintf(_("cannot open file '%s'"),
                                    filename.c_str()),
                       OSError);
  // and read the file
  return readBinaryFile(ifs, filename);
}

// read binary object from stdin
BinaryObject readFile(bool fromStdin)
{
  // read from stdin
//  if (fromStdin)
  return readBinaryFile(cin, (string)_("<STDIN>"));
}

// write binary object to file
void writeBinaryFile(ostream &ofs, string filename, BinaryObject bnw)
{
  // well just dump the data
  ofs.write((char*)bnw._data, bnw._size);
}

// write binary object
void writeFile(string filename, BinaryObject obj)
{
  // open the file
  ofstream ofs(filename.c_str());
  if (ofs.bad())
    throw GsmException(stringPrintf(_("cannot open file '%s'"),
                                    filename.c_str()),
                       OSError);
  // and read the file
  writeBinaryFile(ofs, filename, obj);
}

// write binary object to stdout
void writeFile(bool toStdout, BinaryObject obj)
{
//  if (toStdout)
  writeBinaryFile(cout, (string)_("<STDIN>"), obj);
}

// *** main program

int main(int argc, char *argv[])
{
  try
  {
    // handle command line options
    string destination;
    string source;
    string baudrate;
    string type;
    string subtype;
    int subtypeN;
    bool verbose = false;
    string initString = DEFAULT_INIT_STRING;
    bool swHandshake = false;
    Ref<SieMe> sourceMeTa, destMeTa;
    BinaryObject sourceObject;

    int opt;
    int dummy;
    while((opt = getopt_long(argc, argv, "XI:s:d:b:hvVt:i:", longOpts,
                             &dummy))
          != -1)
      switch (opt)
      {
      case 'X':
        swHandshake = true;
        break;
      case 'I':
        initString = optarg;
        break;
      case 'V':
        verbose = true;
        break;
      case 't':
        type = optarg;
        break;
      case 'i':
        subtype = optarg;
	subtypeN = atoi(optarg);
        break;
      case 'd':
        destination = optarg;
        break;
      case 's':
        source = optarg;
        break;
      case 'b':
        baudrate = optarg;
        break;
      case 'v':
        cerr << argv[0] << stringPrintf(_(": version %s [compiled %s]"),
                                        VERSION, __DATE__) << endl;
        exit(0);
        break;
      case 'h':
        cerr << argv[0] << _(": [-b baudrate][-c][-d device or file][-h]"
                             "[-I init string]\n"
                             "  [-p phonebook name][-s device or file]"
                             "[-t charset][-v]"
                             "[-V][-y][-X]") << endl
             << endl
             << _("  -b, --baudrate    baudrate to use for device "
                  "(default: 38400)")
             << endl
             << _("  -c, --copy        copy source entries to destination")
             << endl
             << _("  -d, --destination sets the destination device to "
                  "connect \n"
                  "                    to, or the file to write") << endl
             << _("  -D, --destination-backend sets the destination backend")
             << endl
             << _("  -h, --help        prints this message") << endl
             << _("  -i, --index       takes index positions into account")
             << endl
             << _("  -I, --init        device AT init sequence") << endl
             << _("  -p, --phonebook   name of phonebook to use") << endl
             << _("  -s, --source      sets the source device to connect to,\n"
                  "                    or the file to read") << endl
             << _("  -t, --charset     sets the character set to use for\n"
                  "                    phonebook entries") << endl
             << _("  -S, --source-backend sets the source backend")
             << endl
             << _("  -v, --version     prints version and exits") << endl
             << _("  -V, --verbose     print detailed progress messages")
             << endl
             << _("  -X, --xonxoff     switch on software handshake") << endl
             << _("  -y, --synchronize synchronize destination with source\n"
                  "                    entries (destination is overwritten)\n"
                  "                    (see gsmpb(1) for details)")
             << endl << endl;
        exit(0);
        break;
      case '?':
        throw GsmException(_("unknown option"), ParameterError);
        break;
      }

    // check if all parameters all present
    if (destination == "" || source == "")
      throw GsmException(_("both source and destination must be given"),
                         ParameterError);

    // start accessing source mobile phone or file
    if (source == "-")
      sourceObject = readFile(true);
    else if (isFile(source))
      sourceObject = readFile(source);
    else
    {
      if (type == "")
        throw GsmException(_("type be given"), ParameterError);
      if (subtype == "")
        throw GsmException(_("subtype be given"), ParameterError);

      sourceMeTa = new SieMe(new
#ifdef WIN32
                            Win32SerialPort
#else
                            UnixSerialPort
#endif
                            (source,
                             baudrate == "" ? DEFAULT_BAUD_RATE :
                             baudRateStrToSpeed(baudrate), initString,
                             swHandshake));
      sourceObject = sourceMeTa->getBinary(type, subtypeN);
    }

    // make sure destination.c_str file exists
    if (destination != "")
    {
      try
      {
        ofstream f(destination.c_str(), ios::out | ios::app);
      }
      catch (exception)
      {
      }
    }

    // start accessing destination mobile phone or file
    if (destination == "-")
      writeFile(true, sourceObject);
    else if (isFile(destination))
      writeFile(destination, sourceObject);
    else
    {
      if (type == "")
        throw GsmException(_("type must be given"), ParameterError);
      if (subtype == "")
        throw GsmException(_("subtype must be given"), ParameterError);

      destMeTa = new SieMe(new 
#ifdef WIN32
                          Win32SerialPort
#else
                          UnixSerialPort
#endif
                          (destination,
                           baudrate == "" ? DEFAULT_BAUD_RATE :
                           baudRateStrToSpeed(baudrate), initString,
                           swHandshake));
      destMeTa->setBinary(type, subtypeN, sourceObject);
    }
  }
  catch (GsmException &ge)
  {
    cerr << argv[0] << _("[ERROR]: ") << ge.what() << endl;
    return 1;
  }
  return 0;
}
